### ISSUE TITLE
Add halo-swap functionality.

### ISSUE THREAD

**Issue by arporter** [stfc/dl_esm_inf#7]
In this issue we will design and implement the necessary interfaces for halo operations; checking whether they are clean and performing halo exchanges.

**Comment by arporter** [comment]
If we follow the example of the LFric infrastructure API (which has no public documentation as yet) then we will need something like:

    type(r2d_field) :: sshn_u

    if(sshn_u%is_dirty(depth=idepth))then
       sshn_u%halo_exch(depth=idepth)
    end if
    ...
    sshn_u%set_clean(to_depth=idepth)
    sshn_u%set_dirty(from_depth=idepth)

**Comment by arporter** [comment]
The above assumed that calling `halo_exch` performs all necessary exchanges to update all halos for a given field out to a given depth. In reality, this is often not required - a kernel may only read from e.g. `sshn_u(ji,jj)` and `sshn_u(ji-1,jj)` and thus only need the left/western halo for this field updated. This is an optimisation but in all probability we will want our interface to support it in the future.

**Comment by arporter** [comment]
Something like:

    integer :: depth
    call sshn_u%halo_exch(mask=depth*W + depth*NW)

where:

    integer, dimension(-1:1,-1:1) :: W, NW
    W=reshape([0,0,0,1,0,0,0,0,0], shape(W))
    NW=reshape([0,0,0,0,0,0,1,0,0], shape(W))

**Comment by arporter** [comment]
`mask` could refer to which neighbours that halo values are required from. However, that exposes information about the domain decomposition in a place where it is not required - the interface should be independent of the domain decomposition being used. A more general approach could be to re-name `mask` as `accesses` and use it to specify the points in the halo that are about to be read and therefore must be updated. It is then up to the infrastructure to decide which processors must provide that information.

**Comment by arporter** [comment]
That last approach is not totally flexible if we have halos with depth greater than unity, e.g.:

    4  x  x  x  x
    3  x  x  x  x
    2  o  o  x  x
    1  o  o  x  x
       1  2  3  4

If we consider the (contrived) example of a case that needs to read from the halo at points (4,3) and (3,4) then there's no way of specifying that with the `accesses` type argument. If this type of situation ever arises (and we wish to optimise for it) then we could add a further argument type where a list of coordinates is provided, e.g.:

    call sshn_u%halo_exch(halo_pts=[2,1,1,2])

where the points are specified relative to the current, owned grid-point at (0,0)

**Comment by rupertford** [comment]
I don't think the above situation can arise as we don't capture it in the metadata and can only pass on the metadata information, see issue 157 in the PSyclone repo https://github.com/stfc/PSyclone/issues/157

**Comment by arporter** [comment]
I've cut-n-pasted in the halo exchange code from the NEMO dCSE projects. This turned out to be way more complicated that I remember so I've deleted most of it. I've got it compiling but it will plenty of work before it will actually work. I also need to add suitable halo-exchange calls in to the example.

**Comment by arporter** [comment]
If sub-domains:
subdomain[   0](   1:  20)(   1:  20), interior:(   2:  21)(   2:  21) 
subdomain[   1](   1:  20)(  21:  40), interior:(   2:  21)(   2:  21) 
then:
`[0](:,22)` is halo which should receive data from `[1](:,2)`
and:
`[1](:,1)` is halo which should receive data from `[0](:,21)`

**Comment by arporter** [comment]
I've realised that the position of a halo region in a field depends on which grid-point type the field lives on. The current halo-swap infrastructure makes no allowance for this so probably gets it right for T-point fields and wrong for everything else.

**Comment by arporter** [comment]
Although I could setup halo information for each field individually, this will be wasteful because I only actually need to do it for each grid-point type.

**Comment by arporter** [comment]
Allocating C-U field with bounds: (1:108,1: 53)
Internal region is:(2:100,2: 51)
Grid has bounds:  (1:108,1: 53)

Allocating C-V field with bounds: (1:108,1: 53)
Internal region is:(2:101,2: 50)
Grid has bounds:  (1:108,1: 53)

Allocating C-T field with bounds: (1:108,1: 53)
Internal region is:(2:101,2: 51)
Grid has bounds:  (1:108,1: 53)

Allocating C-F field with bounds: (1:108,1: 53)
Internal region is:(2:100,2: 50)

**Comment by arporter** [comment]
Problem seems to be with halo swapping fields on V points in the +/-y (j) direction:

    1: halo swap sshn_v
    1: packing from:  2:  9, 11: 11
    1: packed:      0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00
    1: unpacking to:  2:  9, 12: 12
    1: unpacked:      0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00
    1: halo swap sshn_v done
    1: sshn_v internal region:            2           9           2          10

So should be unpacking to y==11 and packing from y==10.

**Comment by arporter** [comment]
Realised that when library code refers to the e.g. "JPlus" direction, this actually refers to the stencil access (i.e. `field(ji,jj+1)`) that requires the communication rather than the direction of the communication! Need to fix this.

**Comment by arporter** [comment]
With the above revelation I now have the shifts working for sends and receives in the j direction. There remains a mismatch in the istart/stop values for these exchanges.

**Comment by arporter** [comment]
Running the example app in my VM I get a crash:

    Process      1 exch tag        8 assigned flags   1 (  1 busy)
    1: tag   11 ireceiving from    1 data  100
    1: packing from:***:***, 51: 51
    At line 1959 of file parallel_comms_mod.f90
    Fortran runtime error: Index '1709098210' of dimension 1 of array 'b2' above upper bound of 108

Presumably this was working OK the last time I tried it? Yes, it works OK on my desktop (including when building with bounds-checking enabled)?!

**Comment by arporter** [comment]
Sub-domains:
    subdomain[   1](   1: 100)(   1:  49), interior:(   2: 101)(   2:  50) 
    subdomain[   1](   1: 100)(   1:  49), interior:(   2: 101)(   2:  50) 

Reported halo-exchange:

    Halo exchange for u:
     2: tag   12 ireceiving from    0 data  100
     2: packing from:  1:100,  2:  2
     2: packed:      0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00
    Halo exchange for u:
    Process      1 exch tag        8 assigned flags   1 (  1 busy)
     1: tag   11 ireceiving from    1 data  100
     1: packing from:  2:101, 51: 51
     1: packed:      0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00
     2: unpacking to:  1:100,  1:  1
     2: unpacked:      0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00     0.0000E+00

**Comment by arporter** [comment]
So PE 1 packs from 2:101 but PE 2 packs from 1:100.

**Comment by arporter** [comment]
Halo exchanges in +/- y are working now. However exchanges for fields on U points in +/-x are not working. Possibly the 'shifts' that I introduced to get +/-y exchanges working need re-visiting for +/-x. I've added tests to the example application that show the problem. (Possibly these tests should not be in the example but it's the simplest place to put them for the moment.)

**Comment by arporter** [comment]
[1,1]<stdout>:subdomain[   1](   1:  50)(   1:  50), interior:(   2:  51)(   2:  51) 
    [1,1]<stdout>:subdomain[   2](  51: 100)(   1:  50), interior:(   2:  51)(   2:  51) 

    [1,1]<stdout>:ERROR in halo value for u_fld:     1    2   51    2  50002.0  51002.0
    [1,1]<stdout>:ERROR in halo value for u_fld:     1    3   51    3  50003.0  51003.0
    [1,1]<stdout>:ERROR in halo value for u_fld:     1    4   51    4  50004.0  51004.0

**Comment by arporter** [comment]
I've realised that at some point I concluded that it would be best if every field was allocated the same extent, irrespective of the grid-point type. I hadn't then realised that this makes life very easy as there are no offsets to worry about:
![ne_stagger_halos](https://user-images.githubusercontent.com/17409212/54823775-bfed7780-4ca0-11e9-930b-1a1614c22b83.png)
That being so, I've had to change the field initialisation code so that all the internal regions for fields on different grid-point types are set-up the same way. I then also had to make sure that points on internal E/W boundaries are set to be wet.
With those changes, halo exchanges now work in both x and y directions.
Next step is to move to running on 4 ranks and dealing with the corners...
